src/pybind/ceph_argparse.py

```diff
@@ -1154,9 +1154,10 @@ def validate_command(sigdict, args, verbose=False):
         else:
             bestcmds.append(cmd)
 
-    # Sort bestcmds by number of args so we can try shortest first
+    # Sort bestcmds by number of req args so we can try shortest first
     # (relies on a cmdsig being key,val where val is a list of len 1)
-    bestcmds_sorted = sorted(bestcmds, key=lambda c: len(c['sig']))
+    is_req_arg = lambda arg: not hasattr(arg, "req") or arg.req == True
+    bestcmds_sorted = sorted(bestcmds, key=lambda c: len([i for i in c['sig'] if is_req_arg(i)]))
```

Comment:
i think `hasattr(arg, "req")` is always true. and i'd suggest something like:

```py
def grade(cmd):
  # prefer optional arguments over required ones
  sigs = cmd['sig']
  return sum(map(lambda sig: [0.5, 1][sig.req], sigs))
bestcmds_sorted = sorted(bestcmds, key=grade)
```

Reply:
thank you. your patch can't pass following case, see

```
./bin/ceph progress --verbose
bestcmds_sorted: 
[{u'flags': 8,
  u'help': u'Reset progress tracking',
  u'module': u'mgr',
  u'perm': u'rw',
  u'sig': [argdesc(<class 'ceph_argparse.CephPrefix'>, name=prefix, numseen=0, req=True, n=1, prefix=progress),
           argdesc(<class 'ceph_argparse.CephPrefix'>, name=prefix, numseen=0, req=True, n=1, prefix=clear)]},
 {u'flags': 8,
  u'help': u'Show progress of recovery operations',
  u'module': u'mgr',
  u'perm': u'r',
  u'sig': [argdesc(<class 'ceph_argparse.CephPrefix'>, name=prefix, numseen=0, req=True, n=1, prefix=progress),
           argdesc(<class 'ceph_argparse.CephInt'>, name=event_id, numseen=0, req=False, n=1),
           argdesc(<class 'ceph_argparse.CephInt'>, name=test, numseen=0, req=False, n=1)]},
 {u'flags': 8,
  u'help': u'Show machine readable progress information',
  u'module': u'mgr',
  u'perm': u'r',
  u'sig': [argdesc(<class 'ceph_argparse.CephPrefix'>, name=prefix, numseen=0, req=True, n=1, prefix=progress),
           argdesc(<class 'ceph_argparse.CephPrefix'>, name=prefix, numseen=0, req=True, n=1, prefix=json),
           argdesc(<class 'ceph_argparse.CephPrefix'>, name=prefix, numseen=0, req=True, n=1, prefix=test)]}]
no valid command found; 3 closest matches:
progress clear
progress {<int>} {<int>}
progress json test
Error EINVAL: invalid command
```

Reply:
we can make it simpler then
```py
def grade(cmd):
  # prefer optional arguments over required ones
  sigs = cmd['sig']
  return sum(map(lambda sig: sig.req, sigs))
bestcmds_sorted = sorted(bestcmds, key=grade)
```